### PR TITLE
Sync up/down as much as possible, even in the presence of errors.

### PR DESCRIPTION
**Pull request by nickmshelley**

### CODE REVIEW

**Comment by nickmshelley** [comment]
I still have a bit of work to do on this (mainly anonymizing data) so don't merge yet. I just wanted to get it out there for maybe a preliminary review to make sure I didn't miss corner cases and that I have good test coverage.

File: LDSAnnotations/AnnotationStore+Annotation.swift [hunk 1]
```diff
@@ -296,6 +296,15 @@ public extension AnnotationStore {
         }
     }
     
+    /// Returns annotations with uniqueIDs
+    public func annotationsWithIDsIn(_ IDs: [Int64]) -> [Annotation] {
```
Comment (hiltonc) [inline LDSAnnotations/AnnotationStore+Annotation.swift]: What do you think about calling this `annotations(ids:)` or `annotations(idsIn:)`?

File: LDSAnnotations/AnnotationStore+SyncIssue.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+class SyncIssueTable {
+    static let table = Table("sync_issue")
+    static let id = Expression<Int64>("_id")
+    static let serviceID = Expression<String>("service_id")
+    static let message = Expression<String?>("message")
+    static let syncErrorType = Expression<SyncErrorType>("sync_error_type")
+    static let failureDate = Expression<Date>("failure_date")
+    
+    static func fromRow(_ row: Row) -> SyncIssue {
+        return SyncIssue(id: row[id],
```
Comment (hiltonc) [inline LDSAnnotations/AnnotationStore+SyncIssue.swift]: Since when did you start manually splitting lines? 😉

File: LDSAnnotations/AnnotationStore+SyncIssue.swift [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                         syncErrorType: row.get(syncErrorType),
+                         failureDate: row[failureDate])
+    }
+}
+
+// MARK: Public
+
+public extension AnnotationStore {
+    /// Returns all sync issues ordered by failure date
+    public func allSyncIssues() -> [SyncIssue] {
```
Comment (hiltonc) [inline LDSAnnotations/AnnotationStore+SyncIssue.swift]: You could just call this `syncIssues`, which would let you add defaulted filter arguments later. Alas, it would look like a method to sync things then 😢

File: LDSAnnotations/Error.swift [hunk 4]
```diff
@@ -36,6 +30,28 @@ public struct SyncError: Error {
     let type: SyncErrorType
 }
 
+extension SyncError {
+    static func processJSONErrors(_ errorsJSON: [[String: Any]], defaultMessage: String, username: String, payloadJSON: Any) -> [SyncError] {
+        let errors: [SyncError] = errorsJSON.map { errorJSON in
+            let message = errorJSON["msg"] as? String ?? defaultMessage
+            let type: SyncErrorType
```
Comment (hiltonc) [inline LDSAnnotations/Error.swift]: You can extract this into a private extension on SyncErrorType:

```swift
extension SyncErrorType {
    init(message: String) {
        switch message {
        case "Invalid label":
            self = .length
        case "Invalid tags":
            self = .invalidTag
        case "Invalid folders":
            self = .missingNotebook
        default:
            self = .unknown
        }
    }
}
```

File: LDSAnnotations/SyncAnnotationsOperation.swift [hunk 5]
```diff
@@ -64,40 +59,29 @@ class SyncAnnotationsOperation: Procedure, ResultInjection {
...
-                    uploadLinkCount: self.uploadLinkCount,
+                let changes = SyncAnnotationsChanges(uploadServiceIDs: self.uploadServiceIDs,
                     downloadAnnotationCount: self.downloadAnnotationCount,
                     downloadNoteCount: self.downloadNoteCount,
                     downloadBookmarkCount: self.downloadBookmarkCount,
                     downloadHighlightCount: self.downloadHighlightCount,
                     downloadTagCount: self.downloadTagCount,
                     downloadLinkCount: self.downloadLinkCount)
-                completion(.success(notebooksResult: notebooksResult, localSyncAnnotationsDate: localSyncAnnotationsDate, serverSyncAnnotationsDate: serverSyncAnnotationsDate, changes: changes, deserializationErrors: self.deserializationErrors))
+                completion(.success(notebooksResult: notebooksResult, localSyncAnnotationsDate: localSyncAnnotationsDate, serverSyncAnnotationsDate: serverSyncAnnotationsDate, changes: changes, errors: self.nonfatalErrors))
```
Comment (hiltonc) [inline LDSAnnotations/SyncAnnotationsOperation.swift]: Should we carry the name `nonfatalError` into the label in enum case `.success`?

File: LDSAnnotations/SyncAnnotationsOperation.swift [hunk 6]
```diff
@@ -36,20 +36,15 @@ class SyncAnnotationsOperation: Procedure, ResultInjection {
...
+    var uploadServiceIDs = [String]()
     
     var downloadAnnotationCount = 0
     var downloadHighlightCount = 0
     var downloadBookmarkCount = 0
     var downloadNoteCount = 0
     var downloadLinkCount = 0
     var downloadTagCount = 0
-    var deserializationErrors = [Error]()
+    var nonfatalErrors = [Error]()
```
Comment (hiltonc) [inline LDSAnnotations/SyncAnnotationsOperation.swift]: It seems odds to group this (via whitespace) with the download counters above it.

File: LDSAnnotations/SyncAnnotationsOperation.swift [hunk 7]
```diff
@@ -123,43 +107,44 @@ class SyncAnnotationsOperation: Procedure, ResultInjection {
                 if let errorsJSON = syncAnnotationsIds?["errors"] as? [[String: Any]], !errorsJSON.isEmpty {
```
Comment (hiltonc) [inline LDSAnnotations/SyncAnnotationsOperation.swift]: This block of code in response to a successful sync is pretty long. Maybe extract it to a private method?

File: LDSAnnotations/SyncIssue.swift [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+import Foundation
+
+/// A Bookmark.
```
Comment (hiltonc) [inline LDSAnnotations/SyncIssue.swift]: More a sync issue than a bookmark.

File: LDSAnnotations/SyncIssue.swift [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        self.id = id
+        self.serviceID = serviceID
+        self.message = message
+        self.syncErrorType = syncErrorType
+        self.failureDate = failureDate
+    }
+    
+}
+
+public func == (lhs: SyncIssue, rhs: SyncIssue) -> Bool {
```
Comment (hiltonc) [inline LDSAnnotations/SyncIssue.swift]: These days I like to make an extension that conforms to `Equatable`, and then put the `==` implementation in that, like:

```swift
extension SyncIssue: Equatable {
    public static function == (lhs: SyncIssue, rhs: SyncIssue) -> Bool {
        ...
    }
}
```

File: LDSAnnotations/AnnotationStore+Annotation.swift (hunk 1 above)
Comment (nickmshelley) [inline LDSAnnotations/AnnotationStore+Annotation.swift]: I'd rather wait for the Swift compiler to catch up to the suggested naming convention. It's very obnoxious to me that this doesn't compile:

``` swift
func annotations(idsIn: [Int]) -> [Int] {
    return idsIn
}

let annotations = annotations(idsIn: [1, 2]) // Error: Variable used within its own initial value
```

But if you prefer I change it still I will.

File: LDSAnnotations/AnnotationStore+SyncIssue.swift (hunk 2 above)
Comment (nickmshelley) [inline LDSAnnotations/AnnotationStore+SyncIssue.swift]: Since the rest of the methods do and I match the style :).

(Although, to be honest, I did try it on one line first and actually found this more readable. Don't tell anyone!)

File: LDSAnnotations/AnnotationStore+SyncIssue.swift (hunk 3 above)
Comment (nickmshelley) [inline LDSAnnotations/AnnotationStore+SyncIssue.swift]: So does that mean you do or do not want it changed? :)

File: LDSAnnotations/SyncIssue.swift (hunk 8 above)
Comment (nickmshelley) [inline LDSAnnotations/SyncIssue.swift]: Heh good catch.

File: LDSAnnotations/AnnotationStore+Annotation.swift (hunk 1 above)
Comment (hiltonc) [inline LDSAnnotations/AnnotationStore+Annotation.swift]: Oh yeah, that's annoying. Nevermind.

File: LDSAnnotations/AnnotationStore+SyncIssue.swift (hunk 2 above)
Comment (hiltonc) [inline LDSAnnotations/AnnotationStore+SyncIssue.swift]: Our little secret.

File: LDSAnnotations/AnnotationStore+SyncIssue.swift (hunk 3 above)
Comment (hiltonc) [inline LDSAnnotations/AnnotationStore+SyncIssue.swift]: Meh :)

**Comment by nickmshelley** [comment]
Ok I think this is ready to review/merge (once tests pass again).

**Comment by stephanheilner** [comment]
I don't see anything that gives me heart burn. There are a lot of changes here though, but overall, I like what you're doing.

File: LDSAnnotations/SyncAnnotationsOperation.swift (hunk 5 above)
Comment (nickmshelley) [inline LDSAnnotations/SyncAnnotationsOperation.swift]: Oops, looks like I forgot this one. I changed _something_... must have been a different result.

**Comment by digbads** [comment]
👍
